### PR TITLE
MESOS: handle graceful termination of pods on task lost status from executor

### LINKED ISSUE #523: Pods stay in graceful termination on executor restart
It was encountered during below scenario:
1. pod scheduled on to a slave and running.
2. mesos-slave (mesos-0.23 with slave recovery) process on that slave was restarted. killing mesos-slave killed the minion/executor/kubelet also. docker container(s) still running.
3. mesos-slave on start tries to recover the framework and executor, but fails to do so and issues kill on the docker container(s) belonging to the pod. slave sends task_lost status
4. scheduler sees task_lost status and reconciles the pod (deletes it). the default gracefulTerminationPeriod is 30secs.

/xref https://github.com/mesosphere/kubernetes-mesos/issues/495

### PR DESCRIPTION
**Pull request by ravilr**
fixes https://github.com/mesosphere/kubernetes-mesos/issues/523

@sttts @jdef

### CODE REVIEW

File: contrib/mesos/pkg/scheduler/scheduler.go [hunk 1]
```diff
@@ -446,7 +447,7 @@ func (k *KubernetesScheduler) reconcileTerminalTask(driver bindings.SchedulerDri
 		// TODO(jdef) for case #2 don't delete the pod, just update it's status to Failed
 		pod := &task.Pod
 		log.Warningf("deleting rogue pod %v/%v for lost task %v", pod.Namespace, pod.Name, task.ID)
-		if err := k.client.Pods(pod.Namespace).Delete(pod.Name, nil); err != nil && !errors.IsNotFound(err) {
+		if err := k.client.Pods(pod.Namespace).Delete(pod.Name, api.NewDeleteOptions(0)); err != nil && !errors.IsNotFound(err) {
```
Comment (sttts) [inline contrib/mesos/pkg/scheduler/scheduler.go]: :+1:

File: contrib/mesos/pkg/scheduler/scheduler.go [hunk 2]
```diff
@@ -437,7 +437,8 @@ func (k *KubernetesScheduler) reconcileTerminalTask(driver bindings.SchedulerDri
 	if (state == podtask.StateRunning || state == podtask.StatePending) &&
 		((taskStatus.GetSource() == mesos.TaskStatus_SOURCE_MASTER && taskStatus.GetReason() == mesos.TaskStatus_REASON_RECONCILIATION) ||
 			(taskStatus.GetSource() == mesos.TaskStatus_SOURCE_SLAVE && taskStatus.GetReason() == mesos.TaskStatus_REASON_EXECUTOR_TERMINATED) ||
-			(taskStatus.GetSource() == mesos.TaskStatus_SOURCE_SLAVE && taskStatus.GetReason() == mesos.TaskStatus_REASON_EXECUTOR_UNREGISTERED)) {
+			(taskStatus.GetSource() == mesos.TaskStatus_SOURCE_SLAVE && taskStatus.GetReason() == mesos.TaskStatus_REASON_EXECUTOR_UNREGISTERED) ||
+			(taskStatus.GetSource() == mesos.TaskStatus_SOURCE_EXECUTOR && taskStatus.GetMessage() == messages.ContainersDisappeared)) {
```
Comment (sttts) [inline contrib/mesos/pkg/scheduler/scheduler.go]: What is the logic behind this? How is this related to graceful termination? If this case was not taken care of, it should have been a problem even before graceful termination was introduced.

**Comment by sttts** [comment]
@ravilr can you split the two changes and add a good explanation for the changes in the commit message?

File: contrib/mesos/pkg/scheduler/scheduler.go (hunk 2 above)
Comment (ravilr) [inline contrib/mesos/pkg/scheduler/scheduler.go]: This is another case of terminal task state status update. It handles task lost status update from executor on docker daemon restart, by deleting the pod so that the RC creates a new one (if the pod is backed by RC). we discussed this in https://github.com/mesosphere/kubernetes-mesos/issues/523

Comment (sttts) [inline contrib/mesos/pkg/scheduler/scheduler.go]: Sounds sensible. Can you add that to the commit message and maybe a short version in the code?

@jdef opinion?

**Comment by ravilr** [comment]
@sttts, splitted into two commits. PTAL.

File: contrib/mesos/pkg/scheduler/scheduler.go (hunk 2 above)
Comment (jdef) [inline contrib/mesos/pkg/scheduler/scheduler.go]: lgtm once documented. at some point we really need to clean this up, but not in this PR

**Comment by ravilr** [comment]
@sttts PTAL

**Comment by sttts** [comment]
@ravilr shippable is not happy. If it won't get green and you think it's a flake, please rebase to latest master.
